pkg/service/auth_context.go

```diff
@@ -23,44 +27,125 @@ type AuthContext struct {
+	Call(ctx common.AuthContext) (interface{}, error)
+}
+
+type configCallback = func(config AuthObjectConfig, obj interface{})
+
+// NewAuthContext creates an AuthContext instance
+func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig config.APIConfig) AuthContext {
+
+	return AuthContext{
+		ParentContext: &parentCtx,
+		Request:       req,
+		API:           &apiConfig,
+		Identity:      make(map[*config.IdentityConfig]interface{}),
+		Metadata:      make(map[*config.MetadataConfig]interface{}),
+		Authorization: make(map[*config.AuthorizationConfig]interface{}),
+	}
+
+}
+
+func (authContext *AuthContext) getAuthObject(ctx context.Context, objConfig AuthObjectConfig, cb configCallback) error {
+	log := ctrl.Log.WithName("Authorino").WithName("AuthContext")
+	select {
+	case <-ctx.Done():
+		log.Info("Context cancelled objConfig %v terminting", objConfig)
+		return nil
+	default:
+		if authObj, err := objConfig.Call(authContext); err != nil {
+			log.Error(err, "Invalid auth object config")
 			return err
 		} else {
-			self.Identity[&c] = ret
-			break
+			cb(objConfig, authObj)
+			return nil
 		}
 	}
+}
+
+func (authContext *AuthContext) getAuthObjects(configs []AuthObjectConfig, cb configCallback) error {
+	log := ctrl.Log.WithName("Authorino").WithName("AuthContext")
+	errGroup, ctx := errgroup.WithContext(context.Background())
+
+	for _, config := range configs {
+		objConfig := config
+		errGroup.Go(func() error {
+			return authContext.getAuthObject(ctx, objConfig, cb)
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		return err
+	} else {
+		log.Info("Successfully fetched all auth objects.", configs)
+		return nil
+	}
+}
+
+// GetIDObject gets an Identity auth object given an Identity config.
+func (authContext *AuthContext) GetIDObject() error {
+	configs := make([]AuthObjectConfig, len(authContext.API.IdentityConfigs))
+	// Convert []SpecificType to []interfaceType
+	for i, conf := range authContext.API.IdentityConfigs {
+		cpConf := conf
+		configs[i] = &cpConf
+	}
+	return authContext.getAuthObjects(configs,
+		func(conf AuthObjectConfig, authObj interface{}) {
+			// Caution: type assertion not checked
+			v, _ := conf.(*config.IdentityConfig)
+			authContext.Identity[v] = authObj
+		})
+}
+
+// GetMDObject gets a Metadata auth object given a Metadata config.
+func (authContext *AuthContext) GetMDObject() error {
+	configs := make([]AuthObjectConfig, len(authContext.API.MetadataConfigs))
+	// Convert []SpecificType to []interfaceType
+	for i, conf := range authContext.API.MetadataConfigs {
+		cpConf := conf
+		configs[i] = &cpConf
+	}
+	return authContext.getAuthObjects(configs,
+		func(conf AuthObjectConfig, authObj interface{}) {
+			// Caution: type assertion not checked
+			v, _ := conf.(*config.MetadataConfig)
+			authContext.Metadata[v] = authObj
+		})
+}
+
+// GetAuthObject gets an Authorization object given an Authorization config.
+func (authContext *AuthContext) GetAuthObject() error {
```

Comment:
```suggestion
func (authContext *AuthContext) EvaluateAuthorization() error {
```

See https://github.com/3scale-labs/authorino/pull/34/files?file-filters%5B%5D=.go#r567739524